runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/StreamingEngineClient.java

```diff
@@ -0,0 +1,477 @@
+        closeStaleStreamsAndCreateNewStreams(ImmutableSet.copyOf(newWindmillConnections.values()));
+    ImmutableMap<Endpoint, Supplier<GetDataStream>> newGlobalDataStreams =
+        createNewGlobalDataStreams(
+            ImmutableSet.copyOf(newWindmillEndpoints.globalDataEndpoints().values()));
+
+    StreamEngineConnectionState newConnectionsState =
+        StreamEngineConnectionState.builder()
+            .setWindmillConnections(newWindmillConnections)
+            .setWindmillStreams(newWindmillStreams)
+            .setGlobalDataEndpoints(newWindmillEndpoints.globalDataEndpoints())
+            .setGlobalDataStreams(newGlobalDataStreams)
+            .build();
+
+    LOG.info(
+        "Setting new connections: {}. Previous connections: {}.",
+        newConnectionsState,
+        connections.get());
+    connections.set(newConnectionsState);
+    isBudgetRefreshPaused.set(false);
+
+    // On first worker metadata. Trigger
+    if (getWorkerMetadataReady.getCount() > 0) {
+      getWorkerMetadataReady.countDown();
+    } else {
+      requestBudgetRefresh(TimeStampedTriggeredBudgetRefresh.Event.NEW_ENDPOINTS);
+    }
+  }
+
+  public ImmutableList<Long> getAndResetThrottleTimes() {
+    StreamEngineConnectionState currentConnections = connections.get();
+
+    ImmutableList<Long> keyedWorkStreamThrottleTimes =
+        currentConnections.windmillStreams().values().stream()
+            .map(WindmillStreamSender::getAndResetThrottleTime)
+            .collect(toImmutableList());
+
+    return ImmutableList.<Long>builder()
+        .add(getWorkerMetadataThrottleTimer.getAndResetThrottleTime())
+        .addAll(keyedWorkStreamThrottleTimes)
+        .build();
+  }
+
+  /** Starts {@link GetWorkerMetadataStream}. */
+  @SuppressWarnings({
+    "FutureReturnValueIgnored", // ignoring Future returned from Executor.submit()
+    "nullness" // Uninitialized value of getWorkerMetadataStream is null.
+  })
+  private void startGetWorkerMetadataStream() {
+    // We only want to set and start this value once.
+    if (getWorkerMetadataStream == null) {
+      synchronized (this) {
+        if (getWorkerMetadataStream == null) {
+          getWorkerMetadataStream =
+              streamingEngineStreamFactory.createGetWorkerMetadataStream(
+                  dispatcherClient.getDispatcherStub(),
+                  getWorkerMetadataThrottleTimer,
+                  endpoints ->
+                      consumeWorkerMetadataExecutor.submit(
+                          () -> consumeWindmillWorkerEndpoints(endpoints)));
+        }
+      }
+    }
+  }
+
+  @SuppressWarnings("FutureReturnValueIgnored")
+  private void startBudgetRefreshThreads() {
+    budgetRefreshExecutor.scheduleAtFixedRate(
+        this::refreshBudget,
+        SCHEDULED_BUDGET_REFRESH_MILLIS,
+        SCHEDULED_BUDGET_REFRESH_MILLIS,
+        TimeUnit.MILLISECONDS);
+  }
+
+  private void refreshBudget() {
```

Comment:
how about this @scwhittle 
Changed to 1 thread that gets scheduled at a fixed internal with an initial delay

This will poll a BlockDeque for a triggered budget refresh without blocking, if a there was a trigger that happened after the most recent refresh, it will redistribute the budget.

Else if enough time has passed and budget has not been distributed, it will redistribute the budget (this is kind of like the scheduling mechanism) 

if lgty, i will go forward and modify the tests :)

Reply:
It seems like if we wanted to trigger budget refresh this would wait to notice until the next poll?  Can we instead just have a thread hanging on consuming from the queue with a timeout? Then it would react immediately to triggering or at most redistribute every timeout period.

This still has the issue though that it will cause things adding to the queue to block if the background thread is busy redistributing budget.  It would be nice if triggering was non-blocking and that multiple signals to trigger were collapsed.  This is what I was wondering if you could do with the Monitor class unless there was some other suitable class. You might be able to use AdvancingPhaser beam class. To trigger call arrive(), the monitoring loop woudl be something like:

int phase = 0;
while (true) {
  phase = phaser.awaitAdvanceInterruptibly(phase, 100, milliseconds);
  if (phase < 0) break; // phaser shutdown
  // either timed out or was triggered, run rebudgeting
}


awaitAdvanceInterruptibly
--


[awaitAdvanceInterruptibly](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/Phaser.html#awaitAdvanceInterruptibly-int-)

Reply:
changed to use phaser.
instead of using a `while(true)` opted to use the ScheduledExecturorService to run each iteration consecutively one after the other with no delay between successive tasks. 

we wait for the scheduled interval or phaser.arrive and then redistribute the budget.

thanks for the rec to use Phaser!
@scwhittle

Reply:
What is the benefit of the fixedRate versus a loop if its running continuously?  I think we still want some way to exit early on shutdown so we don't have to wait for the SCHEDULED_BUDGET_REFRESH for the method to finish.

Reply:
my thought was that if there was an issue/error, the task could just terminate the executor could schedule a new one.

however if this is not a concern we can just use the loop.